### PR TITLE
Solved the task

### PR DESCRIPTION
**Pull request by Likh-Alex**
- Implemented methods `changeEngine`, `changeColor`, `addWheel`
- Created methods `checkEngine` , `clonedWheels`

### CODE REVIEW

File: src/main/java/core/basesyntax/Car.java [hunk 1]
```diff
@@ -1,13 +1,93 @@
...
+        List<Wheel> clonedWheels = new ArrayList<>();
+        for (Wheel wheel : wheels) {
+            clonedWheels.add(wheel.clone());
+        }
+        return clonedWheels;
+    }
+
+    public List<Wheel> getWheels() {
+        return new ArrayList<>(clonedWheels(wheels));
+    }
```
Comment (Ketch92) [inline src/main/java/core/basesyntax/Car.java:55]: You can merge these two methods as they do the same thing

File: src/main/java/core/basesyntax/Car.java [hunk 2]
```diff
@@ -1,13 +1,93 @@
...
+
+    private boolean checkEngine(Engine engine) {
+        return engine != null;
+    }
```
Comment (Ketch92) [inline src/main/java/core/basesyntax/Car.java:59]: I think this method is redundant, you can use the ternary operator for such a small thing

File: src/main/java/core/basesyntax/Car.java [hunk 3]
```diff
@@ -1,13 +1,93 @@
...
+        this.engine = checkEngine(engine) ? engine.clone() : null;
+    }
+
+    public Car changeEngine(Engine engine) {
+        return new Car(year, color, wheels, engine.clone());
+    }
+
+    public Car changeColor(String newColor) {
+        return new Car(year, newColor, wheels, engine.clone());
+    }
```
Comment (Ketch92) [inline src/main/java/core/basesyntax/Car.java:26]: Do you really need to clone here something?

File: src/main/java/core/basesyntax/Car.java [hunk 4]
```diff
@@ -1,13 +1,93 @@
...
+
+    public Car addWheel(Wheel newWheel) {
+        List<Wheel> newWheels = new ArrayList<>();
+        for (Wheel wheel : wheels) {
+            newWheels.add(wheel.clone());
+        }
+        newWheels.add(newWheel);
+        return new Car(year, color, newWheels, engine.clone());
```
Comment (Ketch92) [inline src/main/java/core/basesyntax/Car.java:34]: don't you have a separate method for cloning the wheels?

**Review by Ketch92 — CHANGES_REQUESTED** [review]

**Review by Likh-Alex — COMMENTED** [review]
accidental push of review

File: src/main/java/core/basesyntax/Car.java (hunk 2 above)
Comment (Likh-Alex) [reply src/main/java/core/basesyntax/Car.java]: Done, I agree - ternary operator is a clearer statement

File: src/main/java/core/basesyntax/Car.java (hunk 3 above)
Comment (Likh-Alex) [reply src/main/java/core/basesyntax/Car.java]: Nope, thank you for an idea of doubling the cloning)

File: src/main/java/core/basesyntax/Car.java (hunk 4 above)
Comment (Likh-Alex) [reply src/main/java/core/basesyntax/Car.java]: Refactored the `addWheel` to work with `clonedWheels`

**Review by Ketch92 — APPROVED** [review]

File: src/main/java/core/basesyntax/Car.java [hunk 5]
```diff
@@ -1,13 +1,86 @@
...
+    private final int year;
+    private final String color;
+    private final List<Wheel> wheels;
+    private final Engine engine;
+
+    public Car(int year, String color, List<Wheel> wheels, Engine engine) {
+        this.year = year;
+        this.color = color;
+        this.wheels = clonedWheels(wheels);
+        this.engine = engine != null ? engine.clone() : null;
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/Car.java:17]: `engine != null ? engine.clone() : null;` let’s move this logic into the private method as well

File: src/main/java/core/basesyntax/Car.java [hunk 6]
```diff
@@ -1,13 +1,86 @@
...
+        List<Wheel> clonedWheels = new ArrayList<>();
+        for (Wheel wheel : wheels) {
+            clonedWheels.add(wheel.clone());
+        }
+        return clonedWheels;
+    }
+
+    public List<Wheel> getWheels() {
+        return new ArrayList<>(wheels);
+    }
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/Car.java:52]: should we return deep copy here?

File: src/main/java/core/basesyntax/Car.java [hunk 7]
```diff
@@ -1,13 +1,86 @@
...
+
+    public int getYear() {
+        return year;
+    }
+
+    public String getColor() {
+        return color;
+    }
+
+    private List<Wheel> clonedWheels(List<Wheel> wheels) {
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/Car.java:42]:
```suggestion
    private List<Wheel> getClonedWheels(List<Wheel> wheels) {
```

File: src/main/java/core/basesyntax/Car.java [hunk 8]
```diff
@@ -1,13 +1,86 @@
...
+        List<Wheel> clonedWheels = new ArrayList<>();
+        for (Wheel wheel : wheels) {
+            clonedWheels.add(wheel.clone());
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/Car.java:45]: what if `wheel == null`? let’s add `null` to the new list then

File: src/main/java/core/basesyntax/Car.java [hunk 9]
```diff
@@ -1,13 +1,86 @@
...
+        return Objects.hash(year, color, wheels, engine);
+    }
+
+    @Override
+    protected Car clone() {
+        try {
+            return (Car) super.clone();
+        } catch (CloneNotSupportedException e) {
+            throw new RuntimeException("Can not clone car", e);
+        }
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/Car.java:84]: let's use our constructor here instead

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]

File: src/main/java/core/basesyntax/Car.java (hunk 5 above)
Comment (Likh-Alex) [reply src/main/java/core/basesyntax/Car.java]: done, created `checkIfEngineNull` method

File: src/main/java/core/basesyntax/Car.java (hunk 6 above)
Comment (Likh-Alex) [reply src/main/java/core/basesyntax/Car.java]: done, initialising AL with `clonedWheels`

File: src/main/java/core/basesyntax/Car.java (hunk 7 above)
Comment (Likh-Alex) [reply src/main/java/core/basesyntax/Car.java]: Done

File: src/main/java/core/basesyntax/Car.java (hunk 9 above)
Comment (Likh-Alex) [reply src/main/java/core/basesyntax/Car.java]: simplified by returning new `Car` object

File: src/main/java/core/basesyntax/Car.java (hunk 8 above)
Comment (Likh-Alex) [reply src/main/java/core/basesyntax/Car.java:48]: Done, if any of `wheel` is `null` --> adding as well

File: src/main/java/core/basesyntax/Car.java [hunk 10]
```diff
@@ -1,13 +1,89 @@
...
+        return color;
+    }
+
+    private List<Wheel> getClonedWheels(List<Wheel> wheels) {
+        List<Wheel> clonedWheels = new ArrayList<>();
+        for (Wheel wheel : wheels) {
+            if (wheel == null) {
+                clonedWheels.add(wheel);
+            }
+            clonedWheels.add(wheel.clone());
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/Car.java:48]: line 48 is executed even if wheel == null

File: src/main/java/core/basesyntax/Car.java [hunk 11]
```diff
@@ -1,13 +1,89 @@
...
+        }
+        return clonedWheels;
+    }
+
+    public List<Wheel> getWheels() {
+        return new ArrayList<>(getClonedWheels(wheels));
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/Car.java:54]: do you need to create `new ArrayList<>()` here?

also, let's move private method below the place where we call it

File: src/main/java/core/basesyntax/Car.java [hunk 12]
```diff
@@ -1,13 +1,89 @@
...
+    }
+
+    private Engine checkIfEngineNull(Engine engine) {
+        return engine != null ? engine.clone() : null;
+    }
+
+    public Engine getEngine() {
+        return checkIfEngineNull(engine);
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/Car.java:62]: same here - move the private method below this call

File: src/main/java/core/basesyntax/Car.java [hunk 13]
```diff
@@ -1,13 +1,89 @@
...
+            clonedWheels.add(wheel.clone());
+        }
+        return clonedWheels;
+    }
+
+    public List<Wheel> getWheels() {
+        return new ArrayList<>(getClonedWheels(wheels));
+    }
+
+    private Engine checkIfEngineNull(Engine engine) {
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/Car.java:57]:
```suggestion
    private Engine cloneEngine(Engine engine) {
```

File: src/main/java/core/basesyntax/Car.java [hunk 14]
```diff
@@ -1,13 +1,89 @@
...
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(year, color, wheels, engine);
+    }
+
+    @Override
+    protected Car clone() {
+        return new Car(year, color, getClonedWheels(wheels), engine.clone());
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/Car.java:87]: do you need to pass clones to the constructor? what will happen in the constructor?

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]

File: src/main/java/core/basesyntax/Car.java (hunk 10 above)
Comment (Likh-Alex) [reply src/main/java/core/basesyntax/Car.java]: Added `else` block

File: src/main/java/core/basesyntax/Car.java (hunk 11 above)
Comment (Likh-Alex) [reply src/main/java/core/basesyntax/Car.java]: refactored to return result of `getClonedWheels` itself

Comment (Likh-Alex) [reply src/main/java/core/basesyntax/Car.java]: Moved related methods down below their callers

File: src/main/java/core/basesyntax/Car.java (hunk 12 above)
Comment (Likh-Alex) [reply src/main/java/core/basesyntax/Car.java]: Done

File: src/main/java/core/basesyntax/Car.java (hunk 13 above)
Comment (Likh-Alex) [reply src/main/java/core/basesyntax/Car.java]: Done

File: src/main/java/core/basesyntax/Car.java (hunk 14 above)
Comment (Likh-Alex) [reply src/main/java/core/basesyntax/Car.java]: passing fields directly to the constructor instead of clones.

**Review by KseniiaMakarova — APPROVED** [review]

**Pull request closed without merging**